### ISSUE TITLE
Agent lost connection with the server after cert update

### ISSUE THREAD

**Issue by treicadani** [Ylianst/MeshCentral#412]
Hello,

I am using Meshcentral behind Nginx reverse proxy using company domain with it's installed wild card certificate, and we had a few tens of agents connected for testing, but the ssl certificate expired and we had to renew it, and after we renewed the certificate all the agents lost connection with the server permanently and we had to reinstall all the computers using a newly generated agent. 

Is there a way to get around this? With some unique server generated hash/encryption to create the agent - server identifier, rather than using the server key to generate it?

Else is not an easy task to re-install, re-assign to groups, re-write computer names and the description every time we renew the ssl certificate, on a few hundread computers if we plan to roll out Meshcentral across all devices.

**Comment by Ylianst** [comment]
Hi. MeshCentral is especially designed to handle this situation. When the agent connects to MeshCentral it will keep in mind the certificate from Nginx but keep connecting to MeshCentral and perform a secondary certificate check to a different, long-standing MeshCentral certificate.

The problem your facing is that MeshCentral needs to know what your Ngnix certificate is, so that it is ok allowing Ngnix to "man-in-the-middle" the connection. The dumb & easy solution to get your server running again is to add "ignoreagenthashcheck":true to the "settings" section of the config.json and restart the server, like this:

```
{
  "settings": {
    "Port": 443,
    "IgnoreAgentHashCheck": true
  }
}
```

This will tell MeshCentral to allow any man-in-the-middle TLS connection. This is bad for security. The best is to tell your MeshCentral server where to get the Ngnix certificate like this by adding "CertUrl" in the domain section of config.json like this:

`"CertUrl":"https://myserver.domain.com:443/"`

If you already have this, just restart your MeshCentral server and it will grab the new certificate and you will be ok.

One problem in MeshCentral that I need to fix if that, if Ngnix change the cert all of a sudden, there is no code in MeshCentral to notice all the agents are now seeing a different certificate and auto-reload the Ngnix cert.

In conclusion, if you have CertUrl in your config.json now, just restart your MeshCentral.

Let me know if that works,
Ylian

**Comment by treicadani** [comment]
Hi Ylian, 

I do understand now, i did have CertUrl configured as i configured it by the docs, but there was no mention to restart the server for the changes to take effect if the nginx certificate has changed.

I will test it later on today by setting a dummy cert for nginx then restart the server and see if the agents reconnect and i'll update the result.

Thanks,
Dani

**Comment by hunter-nl** [comment]
I use HAProxy as reverse proxy. CertUrl does not work then. Because HAproxy does not allow connection until MeshCentral server is up and running. CertUrl just waits forever until HAproxy allows to access to MeshCentral web service is up.
So I do it little different. HAproxy gets a new certificate and put this also on the MeshCentral server and do a restart. But then I had the issue the ServerID hash has been changed and then the agents still not connecting to MeshCentral server and need to manually replace ServerID hash in the MeshCentral.msh files on all the agents.
IgnoreAgentHashCheck is bad security so I do not like this setting to be active.

**Comment by treicadani** [comment]
I have tested it with some fake cert and then back to the real cert and got the agents to connect back by restarting Meshcentral.
